src/controllers/auth.ts

```diff
@@ -0,0 +1,108 @@
+import type * as express from 'express'
+
+import { randomBytes } from 'node:crypto'
+import { Get, Produces, Query, Request, Route, SuccessResponse } from 'tsoa'
+import { inject, injectable, singleton } from 'tsyringe'
+
+import { Env } from '../env.js'
+import { ForbiddenError, InternalError } from '../errors.js'
+import { Logger, type ILogger } from '../logger.js'
+import IDPService from '../models/idpService.js'
+import { HTMLController } from './HTMLController.js'
+
+function base64URLEncode(buf: Buffer) {
+  return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '')
+}
+
+const nonceCookieOpts: express.CookieOptions = {
+  sameSite: true,
+  httpOnly: true,
+  signed: true,
+  secure: true,
+  path: '/auth/redirect',
+}
+
+const tokenCookieOpts: express.CookieOptions = {
+  sameSite: true,
+  maxAge: 24 * 60 * 60 * 1000,
+  httpOnly: true,
+  signed: true,
+  secure: true,
+}
+
+@singleton()
+@injectable()
+@Route('/auth')
+@Produces('text/html')
+export class AuthController extends HTMLController {
+  private redirectUrl: string
+  private logger: ILogger
+
+  constructor(
+    private env: Env,
+    private idp: IDPService,
+    @inject(Logger) logger: ILogger
+  ) {
+    super()
+    this.logger = logger.child({ controller: '/auth' })
+    this.redirectUrl = `${env.get('PUBLIC_URL')}/auth/redirect`
+  }
+
+  /**
+   * Retrieves the root page for the site
+   */
+  @Get('/login')
+  @SuccessResponse(302, 'Redirect')
+  public async login(@Request() req: express.Request, @Query() path: string): Promise<void> {
+    const { res } = req
+    if (!res) {
+      throw new InternalError()
+    }
+
+    // make random state
+    const nonce = base64URLEncode(randomBytes(32))
+    res.cookie('VERITABLE_NONCE', nonce, nonceCookieOpts)
+
+    // setup for final redirect
+    res.cookie('VERITABLE_REDIRECT', path, nonceCookieOpts)
+
+    const redirect = new URL(await this.idp.authorizationEndpoint)
+    redirect.search = new URLSearchParams({
+      response_type: 'code',
+      client_id: this.env.get('IDP_CLIENT_ID'),
+      redirect_uri: this.redirectUrl,
+      state: nonce,
+      scope: 'openid',
+    }).toString()
+
+    this.logger.debug('login redirect to %s', redirect)
+    res.redirect(302, redirect.toString())
+  }
+
+  @Get('/redirect')
+  @SuccessResponse(302, 'Redirect')
```

Comment:
got to say a compliment for status code usage, you've been on top as far as I can remember and this is maybe first time I'm saying (hopefully)